### PR TITLE
Concurrency refactors and cleanups

### PR DESCRIPTION
**Pull request by MisterDA**
While working on replacing winpthreads (the pthreads emulation layer on Windows) with direct use of the WinAPI concurrency primitives, I've gathered a few refactors and cleanups on systhreads and the runtime.
The end goal is to share common code between platforms using pthreads and Windows, and have platform specific code in different files. As it doesn't make much sense to have `st_pthreads.h` and `st_posix.h` I'm removing `st_pthreads.h` and inlining relevant code in either `st_posix.h` or `st_win32.h`, and introducing a new `st_stubs.h` for common code.
I'm opening this PR now because it mildly conflicts with #12410 and #13348, and I don't enjoy dealing with conflicts too much (but yay for `git rerere`), so better now than later? (of course if they get merged first, I'll rebase this PR).

### CODE REVIEW

**Comment by gasche** [comment]
I'm confused by this PR.

- Some commits are called "removed unexposed and unused `foo_retcode` typedef", but in fact the type *is* used in the codebase, apparently to track (as annotations) values that should be passed to `sync_check_error`. I would intuitively assume that the people who wrote this believed that this abstraction improves the code / reduces the chance for mistakes / makes it easier to check that the codebase is consistently checking for errors. Do you believe that this assumption does not hold anymore, and why? The only argument you hint at is that this typedef is not exposed in the user-facing APIs, I don't understand how that justifies removing it.

- The last commit ("Inline st_pthreads.h into st_posix.h and st_win32.h") appears to duplicate a large amount of code, including documentation comments, etc. (github reports it at 594 added lines for 386 removed lines). From a distance this looks like the opposite of a desirable refactoring to me. What am I missing?

> As it doesn't make much sense to have `st_pthreads.h` and `st_posix.h`

My understanding is that `st_pthreads.h` contains common code that works with either pthreads (on Unixes) or winpthreads (on Windows), and `st_posix.h` contains two less-portable functions that deals with signals and is not implemented on Windows (`st_win32.h` only has placeholder definitions that fail at runtime).

**Comment by gasche** [comment]
Notes on the other commits (skimmed quickly):

- I'm happy with changes to use C99 booleans (I think we for now are trying not to use them in the user-facing API, but using them internally is good), feel free to submit a separate PR about it.

- I'm less happy with the commit that removes `return;` statement at the end of void-returning functions. The people who wrote the code apparently believed that it was good style, and your preference to remove them is just a difference in taste -- as far as I can tell there is no objective software-engineering justification for actively adding or removing `return;` statements at the end of void-returning functions. I don't want to waste our time looking at cosmetic changes that do not actually improve the codebase. Every change has a cost (in review time, rebase work, (not this time) regression risk, etc.), and "I like it better this way" is not enough benefit to offset the cost.

**Comment by MisterDA** [comment]
Thanks a lot for the review!

> Some commits are called "removed unexposed and unused `foo_retcode` typedef", but in fact the type _is_ used in the codebase, apparently to track (as annotations) values that should be passed to `sync_check_error`. I would intuitively assume that the people who wrote this believed that this abstraction improves the code / reduces the chance for mistakes / makes it easier to check that the codebase is consistently checking for errors. Do you believe that this assumption does not hold anymore, and why? The only argument you hint at is that this typedef is not exposed in the user-facing APIs, I don't understand how that justifies removing it.

I think the typedef would have been useful if the C compiler could somehow check conversions between the type alias and the type behind the typedef, a bit like private types in OCaml. It would also have been useful if we were declaring a custom error type. The `caml_check_error` or `sync_check_error` are not using the typedef. Reading the code I would have preferred for it to call a spade a spade and not rename the error type. If anything, maybe we didn't go far enough in the use of this abstraction.

> The last commit ("Inline st_pthreads.h into st_posix.h and st_win32.h") appears to duplicate a large amount of code, including documentation comments, etc. (github reports it at 594 added lines for 386 removed lines). From a distance this looks like the opposite of a desirable refactoring to me. What am I missing?

In `st_pthreads.h` there's code using the pthread API, and code using our threads or runtime API. Code using pthreads needs to be ported on Windows, and the other code may be shared. Moving the code around eases the review for the follow-up PR, but I overlooked that this commit is too early for this PR.

> > As it doesn't make much sense to have `st_pthreads.h` and `st_posix.h`
> 
> My understanding is that `st_pthreads.h` contains common code that works with either pthreads (on Unixes) or winpthreads (on Windows), and `st_posix.h` contains two less-portable functions that deals with signals and is not implemented on Windows (`st_win32.h` only has placeholder definitions that fail at runtime).

Thanks for the explanation, I'll hold this commit for a while. It's not too bad to merge `st_pthreads.h` into `st_posix.h`, the "p" stands for "posix".

> I'm less happy with the commit that removes return; statement at the end of void-returning functions.

IMO this is unnecessary code (early returns matter, but these?), ~inconsistent with the rest of the code base,~ and the commit reduces the line count.

**Comment by gasche** [comment]
You removed the commits I was grumpy about (thanks!).

I had a new look and they are overall fine, but I don't know about the last one ("extract common code that is not posix/pthreads"). I find the current (trunk) structure a bit complex already (st_pthreads.h is a "mixin" that both provides features and expects requirements from its includer), and the new layering you propose is even slightly more complex.

Two questions about this:
- One change in that commit is to move the `tick_thread` related code out of `st_pthreads.h`. Could you include it directly in `st_stubs.c`?
- It would be nice to also move the `bt_lock_{acquire,release}` stuff in `st_stubs.c`, but `release` depends on `st_masterlock_waiters(m)` which is in st_pthread.h. Could the prototype `void st_bt_lock_release(st_masterlock *m)` be changed to `void st_bt_lock_release(bool other_threads_waiting)` to cut the dependency cycle?

**Comment by MisterDA** [comment]
> One change in that commit is to move the `tick_thread` related code out of `st_pthreads.h`. Could you include it directly in `st_stubs.c`?

Done.

> It would be nice to also move the `bt_lock_{acquire,release}` stuff in `st_stubs.c`, but `release` depends on `st_masterlock_waiters(m)` which is in `st_pthreads.h`. Could the prototype `void st_bt_lock_release(st_masterlock *m)` be changed to `void st_bt_lock_release(bool other_threads_waiting)` to cut the dependency cycle?

It's also just possible to add prototypes for `bt_lock_{acquire,release}` and `st_masterlock_waiters` in `st_pthreads.h` and put the functions definitions in `st_stubs.c`. Would that be ok with you?

**Comment by gasche** [comment]
> It's also just possible to add prototypes for `bt_lock_{acquire,release}` and `st_masterlock_waiters` in st_pthreads.h and put the functions definitions in st_stubs.c. Would that be ok with you?

I find this a bit complex. Maybe we should do the work of writing down the interface of both files, what they expect from the other file and what they provide?

If this is more work than you want to tackle, personally I would err on the side of keeping things in st_pthreads.h for now, we don't need to decouple everything in one go.

**Comment by MisterDA** [comment]
> Maybe we should do the work of writing down the interface of both files, what they expect from the other file and what they provide?

I'll give it a go.

> If this is more work than you want to tackle, personally I would err on the side of keeping things in st_pthreads.h for now, we don't need to decouple everything in one go.

I'm working on replacing winpthreads with Windows primitives. My idea was to extract functions that are not using pthreads/unix signals from `st_pthreads.h` and `st_posix.h`, so that `st_win32.h` can simply provide the same interface as `st_pthreads.h` and `st_posix.h` without duplicated code.

File: runtime/domain.c
Comment (dra27) [inline runtime/domain.c]:
```suggestion
    caml_check_error(err, "failed to create domain backup thread");
```

File: runtime/domain.c
Comment (dra27) [inline runtime/domain.c]: As above, this can also just call `caml_check_error` without the `if`

File: otherlibs/systhreads/st_stubs.c
Comment (dra27) [inline otherlibs/systhreads/st_stubs.c]: I think this could probably have been removed in 5.0 itself - if it's not needed as part of the subsequent removal, I think it can just go (in 4.14 it had a purpose)

File: otherlibs/systhreads/st_stubs.c
Comment (dra27) [inline otherlibs/systhreads/st_stubs.c:76]: Pedantically, if these were supposed to be moves:
```suggestion

  return;
}
```

File: otherlibs/systhreads/st_stubs.c
Comment (dra27) [inline otherlibs/systhreads/st_stubs.c:90]:
```suggestion

  return;
}
```

**Review by dra27 — APPROVED** [review]
LGTM, once @gasche is also happy

**Comment by MisterDA** [comment]
Thanks for the reviews!

I've moved the declaration of the functions in `st_subs.c` and used a forward declaration of `st_masterlock`. I think the code is a bit clearer this way, I don't have a better idea on how to improve on that.

**Comment by gasche** [comment]
> I don't have a better idea on how to improve on that.

Have you considered the suggestion I made [earlier](https://github.com/ocaml/ocaml/pull/13352#issuecomment-2299316030)? It was as follows:

>  Could the prototype `void st_bt_lock_release(st_masterlock *m)` be changed to `void st_bt_lock_release(bool other_threads_waiting)` to cut the dependency cycle?

**Comment by MisterDA** [comment]
> > I don't have a better idea on how to improve on that.
> 
> Have you considered the suggestion I made [earlier](https://github.com/ocaml/ocaml/pull/13352#issuecomment-2299316030)? It was as follows:
> 
> > Could the prototype `void st_bt_lock_release(st_masterlock *m)` be changed to `void st_bt_lock_release(bool other_threads_waiting)` to cut the dependency cycle?

Indeed, I forgot this. It's really nice now, thanks!

File: runtime/caml/platform.h
Comment (ghost) [inline runtime/caml/platform.h]: What is the rationale for this change? It looks wrong to me.

Comment (MisterDA) [reply runtime/caml/platform.h]: `DEBUG_UNLOCK` decrements a thread local counter that holds the current number of locks held. If it is called before `pthread_mutex_unlock` and the latter fails, the count will have been lowered and there's a fatal error (thanks to `check_err`
If we want to inspect the state of the runtime with this debug counter, I think it makes more sense to decrement it iff `pthread_mutex_unlock` succeeded. 
Or is there something I missed?

Comment (ghost) [reply runtime/caml/platform.h]: But if the unlock succeeds, you're opening a race where another thread can suddenly fail `caml_plat_assert_all_locks_unlocked()` between the unlock and the `DEBUG_UNLOCK` adjustment.

Comment (MisterDA) [reply runtime/caml/platform.h]: You're right. I'll fix my Windows code too.
Unfortunately I don't have access to a computer until next Monday, so I can't remove this patch from the PR.

Comment (ghost) [reply runtime/caml/platform.h]: Noprob - just make sure noone puts a "merge me" label on the PR. Or mark it as draft, which will prevent it from being merged by mistake.

Comment (MisterDA) [reply runtime/caml/platform.h]: But isn't there also a race in the current code? If thread A calls `DEBUG_UNLOCK` and thread B calls the all unlocked assertion before `pthread_unlock` then there's also an inconsistency.

Comment (ghost) [reply runtime/caml/platform.h]: Yes. Welcome to the joys of parallel programming. You can't have a reliable check without support from the OS (or at least, from the threads library). But narrower races are preferrable here.

Comment (MisterDA) [reply runtime/caml/platform.h]: I've removed this change.

**Comment by MisterDA** [comment]
I believe I've addressed all the comments.

**Comment by gasche** [comment]
You forgot to update the Changes entry to credit @dustanddreams' bug-catching review.

**Review by gasche — APPROVED** [review]

**Comment by gasche** [comment]
Merged, thanks!

**Pull request merged**
